os/mm/mm_heap/mm_heapinfo_group.c

```diff
@@ -0,0 +1,189 @@
+ * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ * COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
+ * OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ****************************************************************************/
+/****************************************************************************
+ * Included Files
+ ****************************************************************************/
+#include <tinyara/config.h>
+#include <tinyara/sched.h>
+#include <tinyara/mm/mm.h>
+#include <stdint.h>
+#include <string.h>
+#include <tinyara/mm/heapinfo_internal.h>
+
+int max_group;
+struct heapinfo_group_s heapinfo_group[HEAPINFO_USER_GROUP_NUM];
+struct heapinfo_group_info_s group_info[HEAPINFO_THREAD_NUM];
+
+/****************************************************************************
+ * Name: heapinfo_update_group
+ *
+ * Description:
+ * Update Peak heap size for Group
+ ****************************************************************************/
+void heapinfo_update_group(mmsize_t size, pid_t pid)
+{
+	int check_idx;
+	int group_num;
+	int stack_pid;
+	for (check_idx = 0; check_idx < HEAPINFO_THREAD_NUM; check_idx++) {
+		if (pid == group_info[check_idx].pid) {
+			group_num = group_info[check_idx].group;
+			heapinfo_group[group_num].curr_size += size;
+			/* Update peak size */
+			if (heapinfo_group[group_num].curr_size > heapinfo_group[group_num].peak_size) {
+				heapinfo_group[group_num].peak_size = heapinfo_group[group_num].curr_size;
+				/* calculate the summation of stacks */
+				heapinfo_group[group_num].stack_size = 0;
+				for (stack_pid = 0; stack_pid < HEAPINFO_THREAD_NUM; stack_pid++) {
+					if (group_info[stack_pid].pid != -1 && group_info[stack_pid].group == group_num) {
+						heapinfo_group[group_num].stack_size += group_info[stack_pid].stack_size;
+					}
+				}
+				heapinfo_group[group_num].heap_size = heapinfo_group[group_num].peak_size - heapinfo_group[group_num].stack_size;
+			}
+			break;
+		}
+	}
+}
+
+/****************************************************************************
+ * Name: heapinfo_update_group_info
+ *
+ * Description:
+ * when create or release task/thread, check that the task/thread is 
+ * in group list
+ ****************************************************************************/
+void heapinfo_update_group_info(pid_t pid, int group, int type)
+{
+	int info_idx;
+	struct tcb_s *tcb;
+	switch (type) {
+	case HEAPINFO_INIT_INFO:
+		for (info_idx = 0; info_idx < HEAPINFO_THREAD_NUM; info_idx++) {
+			group_info[info_idx].pid = pid;
```

Comment:
One question it seems that change entire of group_info based on given values in parameter, is it acceptable?
if each of caller call this function with different parameter,  it will changed all the time, is it intended?
I am asking this basically init assign initialize with defined value(something like .. 0 or -1...or some invalid or ideal value) but this is depend on parameter...
That's why I am asking this.

Reply:
`HEAPINFO_INIT_INFO` is called only from mm_initialize. So I think it is okay.

Reply:
I mean it should use specific value instead of values from parameter. because it is init.

Reply:
When calling this function with `HEAPINFO_INIT_INFO`, `pid` will be passed with -1.
I will change the -1 to specific defined value(INVALID_PROCESS_ID).

Reply:
Yes this is callee, so you should use specific value instead of parameter, even if there's no chance that being called from another place, like a HEAPINFO_DEL_INFO in below. Init / deinit should not have a dependency from caller.